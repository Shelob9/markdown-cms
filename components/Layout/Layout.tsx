import React, { useState } from 'react'
import { Provider } from 'react-slot-fill'

import {
	P,
	Box,
	Heading,
	A,
} from './primatives'

import { NextSeo } from 'next-seo'
import Link from 'next/link'
import { useTheme } from '../../ThemeProvider'

export type linksList = { href: string; label: string; hoverText: string }[]


const Menu: React.FC<{ className?: string; withHover?: boolean;links: linksList }> = ({
	className,
    withHover,
    links
}) => {
	const [hovered, setHovered] = useState<string>('')
	const theHoveredLink = links.find((l) => l.href === hovered)

	return (
		<nav className={className}>
			<ul className="list-reset lg:flex justify-end items-center">
				{links.map((link) => (
					<li
						key={link.href}
						className="mr-3 py-2 lg:py-0"
						onMouseEnter={() => setHovered(link.href)}
						onMouseLeave={() => setHovered('')}
					>
						<Link href={link.href}>
							<a
								className="inline-block py-2 px-4 text-gray-900 hover:text-gray-500 hover:border-gray-400 font-bold no-underline dark:text-gray-200"
								title={link.hoverText}
							>
								{link.label}
							</a>
						</Link>
					</li>
				))}
			</ul>
			{withHover && <P> {hovered && <>{theHoveredLink.hoverText}</>}</P>}
		</nav>
	)
}
const Layout = (props: {
	title: string
	displayTitle?: string
	children: any
	innerSize?: 1024 | 728 | undefined
	description?: string
	prose?: boolean
	displaySubtitle?: string
	headerLinks?:linksList
}) => {
	const {
		children,
		description,
		title,
		displayTitle,
		prose,
		displaySubtitle,
		headerLinks
	} = props
	const innerSize = props.innerSize ? props.innerSize : 728
	const { darkMode, toggleDarkMode } = useTheme()
	const [showMenu, setShowMenu] = useState(false)

	return (
		<>
			<Layout.SEO title={title} description={description} />
			<Provider>
				<Box className="min-h-screen">
					<div className={'items-center'}>
						<Layout.Header title={title} links={headerLinks} />
						<div className="container w-full flex flex-wrap mx-auto px-2 pt-8 sm:pt-4 mt-0">
							<div className="p-8 pt-2 mt-6 lg:mt-0 text-gray-900 leading-normal  border border-gray-400 border-rounded min-w-full">
								{displayTitle && (
									<div className="font-sans">
										<h1 className="font-sans break-normal text-gray-900 pt-6 pb-2 text-xl">
											{displayTitle}
										</h1>
										{displaySubtitle && (
											<P>{displaySubtitle}</P>
										)}
										<hr className="border-b border-gray-400" />
									</div>
								)}
								<main
									className={prose ? 'prose prose-2xl' : ''}
								>
									{children}
								</main>
							</div>
						</div>
					
					</div>
				</Box>
			</Provider>
		</>
	)
}

Layout.Header = (props: { title: string,links?:linksList }) => {
    
	function navToggle() {
		var btn = document.getElementById('menuBtn')
		var nav = document.getElementById('menu')
		btn.classList.toggle('open')
		nav.classList.toggle('flex')
		nav.classList.toggle('hidden')
	}
	return (
		<header
			id="top"
			className=" w-full flex flex-col fixed sm:relative bg-white pin-t pin-r pin-l"
		>
			<nav
				id="site-menu"
				className="flex flex-col sm:flex-row w-full justify-between items-center px-4 sm:px-6 py-1 bg-white shadow sm:shadow-none border-t-4 border-red-900"
			>
				<div className="w-full sm:w-auto self-start sm:self-center flex flex-row sm:flex-none flex-no-wrap justify-between items-center">
					<span className="no-underline py-1">
						<h1 className="font-bold text-lg tracking-widest">
							<img
								src={'/icons/icon-72x72.png'}
								className={'logo mr-4'}
								alt={'Logo'}
							/>
							<Link href={'/'}>
								<a className="text-gray-900 dark:text-gray-200  no-underline hover:no-underline font-extrabold text-xl lg:pl-4">
									{props.title}
								</a>
							</Link>
						</h1>
					</span>
					<button
						id="menuBtn"
						className="hamburger block sm:hidden focus:outline-none"
						type="button"
						onClick={() => navToggle()}
					>
						<span className="hamburger__top-bun"></span>
						<span className="hamburger__bottom-bun"></span>
					</button>
				</div>
				<div
					id="menu"
					className="w-full sm:w-auto self-end sm:self-center sm:flex flex-col sm:flex-row items-center h-full py-1 pb-4 sm:py-0 sm:pb-0 hidden"
				>
					{props.links && props.links.map((link) => (
						<Link key={link.href} href={link.href}>
							<a
								className="text-dark font-bold hover:text-red text-lg w-full no-underline sm:w-auto sm:pr-4 py-2 sm:py-1 sm:pt-2"
								title={link.hoverText}
							>
								{link.label}
							</a>
						</Link>
					))}
				</div>
			</nav>
		</header>
	)
}
Layout.SEO = ({ title, description }) => (
	<NextSeo title={title} description={description} />
)

Layout.Footer = (props: {
    About: () => JSX.Element,
    socialLinks?: linksList
}) => {
    const { About,socialLinks } = props;
    return (
        <footer className="bg-white dark:bg-gray-500 border-t border-gray-400 shadow">
            <div className="container mx-auto flex py-8">
                <div className="w-full mx-auto flex flex-wrap">
                    <Box as={'aside'} className="flex w-full lg:w-1/2 ">
                        <Box className="px-8">
                            <About />
                        </Box>
                    </Box>
                    <Box
                        as={'aside'}
                        className="flex w-full lg:w-1/2 lg:justify-end lg:text-right"
                    >
                        <Box className="px-8">
                            <Heading level={3} className="font-bold text-gray-900">
                                Social
						    </Heading>
                            <ul className="list-reset items-center text-sm pt-3">
                                {socialLinks && socialLinks.map(({ href, label }) => {
                                    <li key={href}>
                                        <a
                                            className="inline-block text-gray-600 no-underline hover:text-gray-900 hover:underline py-1"
                                            href={href}
                                        >
                                            {label}
                                        </a>
                                </li>
                                })}
                            </ul>
                        </Box>
                    </Box>
                </div>
            </div>
        </footer>
    )
}

export default Layout
