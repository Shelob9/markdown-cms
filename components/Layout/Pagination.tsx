import React from 'react'
import Link from 'next/link'
type pageLink = {
	href: string
	label: string
}
const Pagination = (props: {
	pageLinks?: pageLink[]
	nextPage?: pageLink
	prevPage?: pageLink
}) => {
	const { pageLinks, nextPage, prevPage } = props
	return (
		<React.Fragment>
			<nav aria-label="pagination">
				<ul className="pagination">
					{prevPage && (
						<li>
							<Link href={prevPage.href}>
								<a>{prevPage.label}</a>
							</Link>
						</li>
					)}
					{pageLinks && (
						<ul>
							{pageLinks.map((pageLink) => (
								<li>
									<Link href={pageLink.href}>
										<a>{pageLink.label}</a>
									</Link>
								</li>
							))}
						</ul>
					)}
					{nextPage && (
						<li>
							<Link href={nextPage.href}>
								<a href={nextPage.href}>{nextPage.label}</a>
							</Link>
						</li>
					)}
				</ul>
			</nav>
		</React.Fragment>
	)
}

export default Pagination
