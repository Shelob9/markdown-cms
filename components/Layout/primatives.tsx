import React, { createElement, CSSProperties } from 'react'
import Link from 'next/link'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'

export const Box: React.FC<{
	as?: string
	className?: string
	children: any
	style?: any | CSSProperties
	id?: string
}> = ({ as, children, className, style, id }) => {
	return createElement(as ? as : 'div', { className, style, id }, children)
}

export const Quote: React.FC<{
	children: any
	cite: string
	RenderCite?: () => JSX.Element
	citeHref?: string
}> = ({ children, cite, citeHref, RenderCite }) => {
	return (
		<Box as={'blockquote'}>
			<Box as={'div'}>{children}</Box>
			{!RenderCite ? (
				<Box as={'cite'}>
					{citeHref ? <a href={citeHref}>{cite}</a> : <>{cite}</>}
				</Box>
			) : (
				<RenderCite />
			)}
		</Box>
	)
}

export const Tabbed: React.FC<{
	tabs: { title: string; key: string; Render: () => JSX.Element }[]
}> = ({ tabs }) => {
	return (
		<Tabs>
			<TabList className={'list-none'}>
				{tabs.map(({ title, key }) => {
					return <Tab key={key}>{title}</Tab>
				})}
			</TabList>
			{tabs.map(({ Render, key }) => {
				return (
					<TabPanel key={key}>
						<Render />
					</TabPanel>
				)
			})}
		</Tabs>
	)
}

export const Button: React.FC<{
	outline?: boolean
	onClick?: () => void
	variant?: 'primary' | 'seconday' | 'default' | 'dark' | 'error'
	children: any
	className?: string
	title?: string
}> = ({ outline, children, onClick, variant, title }) => {
	let className = variant ? variant : ''

	className = outline
		? `${className} ${variant} outline dark:text-white`
		: 'outline dark:text-white m'
	return (
		<button
			className={className}
			onClick={(e) => {
				if (onClick) {
					e.preventDefault()
					onClick()
				}
			}}
			title={title}
		>
			{children}
		</button>
	)
}

export const Heading: React.FC<{
	level?: number
	children: any
	className?: string
}> = ({ level, children, className }) => {
	return createElement(`h${level ? level : 2}`, { className }, children)
}

export const NavLink: React.FC<{
	children: any
	href: string
	className?: string
}> = ({ children, href, className }) => {
	return (
		<A
			className={className ? `${className} nav-item` : 'nav-item'}
			href={href}
		>
			{children}
		</A>
	)
}

export const P: React.FC<{
	children: any
	className?: string
}> = ({ children, className }) => {
	return (
		<Box as={'p'} className={className}>
			{children}
		</Box>
	)
}

export const Div: React.FC<{
	children: any
	className?: string
	style?: any | CSSProperties
}> = ({ children, className, style }) => {
	return (
		<Box as={'div'} className={className} style={style}>
			{children}
		</Box>
	)
    }

export const A: React.FC<{
	children: any
	href: string
	className?: string
}> = ({ children, href, className }) => {
	if (href.startsWith('http')) {
		return (
			<a
				className={className}
				href={href}
				target="_blank"
				rel="noopener noreferrer"
			>
				{children}
			</a>
		)
	}
	return (
		<Link href={href}>
			<a className={className}>{children}</a>
		</Link>
	)
}

export const Divider = () => {
	return <hr />
}

export const Section = (props) => (
	<Box as={'section'} {...props}>
		{props.children}
	</Box>
)

export const Input = (props) => {
	return <input {...props} />
}

export const Label = (props) => (
	<Box as={'label'} {...props}>
		{props.children}
	</Box>
)

export const Grid: React.FC<{ children: any; className?: string }> = ({
	children,
	className,
}) => (
	<Box as={'div'} className={`flex mb-4 ${className}`}>
		{children}
	</Box>
)

export const Column: React.FC<{
	children: any
	width: '1/3' | '1/4' | '1/2' | '3/4' | '2/3' | 'full'
	className?: string
}> = ({ children, width, className }) => (
	<Box as={'div'} className={`w-${width} ${className} sm:w-full`}>
		{children}
	</Box>
)

export const Card: React.FC<{
	Content: () => JSX.Element
	title: string
	Title?: () => JSX.Element
	Footer: () => JSX.Element
	hLevel?: number
}> = ({ Content, title, Footer, hLevel, Title }) => {
	return (
		<Box className="card">
			<header>
				{Title ? (
					<Title />
				) : (
					<Heading level={hLevel ? hLevel : 3}>{title}</Heading>
				)}
			</header>
			<Content />
			<footer>
				<Footer />
			</footer>
		</Box>
	)
}
