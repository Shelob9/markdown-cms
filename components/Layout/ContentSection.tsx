import React from 'react'
import { Section, Box } from './primatives'
const ContentSection = (props: {
	children: any
	title: string
	hLevel?: 1 | 2 | 3 | 4
}) => {
	return (
		<Section className="text-gray-700 body-font overflow-hidden">
			<Box as={'div'}>
				<Box as={'div'}>{''}</Box>
				{React.createElement(
					`h${props.hLevel ? props.hLevel : 3}`,
					{
						className:
							'text-2xl font-medium text-gray-900 title-font mb-2',
					},
					[props.title]
				)}
				{props.children}
			</Box>
		</Section>
	)
}

export default ContentSection
