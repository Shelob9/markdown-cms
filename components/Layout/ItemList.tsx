import Link from 'next/link'
import React, { Fragment } from 'react'
import { Box, Divider, Heading } from './primatives'
export type listItem = {
	id: number | string
	date?: string
	title: string
	to: string
	content: string
}
const Div = (props) => (
	<Box as={'div'} {...props}>
		{props.children}
	</Box>
)

const Section = (props) => (
	<Box as={'section'} {...props}>
		{props.children}
	</Box>
)

export const ItemSingle = (props: { item: listItem; hLevel: number }) => {
	const { item, hLevel } = props
	const date = item.date ? new Date(item.date) : undefined

	const RenderLink = (props: { children: string }) => (
		<Fragment>
			{item.to.startsWith('http') ? (
				<a href={item.to} target="_blank" rel="noopener noreferrer">
					{props.children}
				</a>
			) : (
				<Link href={item.to}>
					<a>{props.children}</a>
				</Link>
			)}
		</Fragment>
	)

	return (
		<Div py={1} m={1} key={item.id} border={1} bg="muted">
			<Box>
				<Heading level={hLevel}>
					<RenderLink>{item.title}</RenderLink>
				</Heading>
				{date && (
					<Box>
						<span>{date.toLocaleDateString()}</span>
					</Box>
				)}
				{item.to && <RenderLink>Read More</RenderLink>}
			</Box>
			<Divider />
		</Div>
	)
}

const ItemList = (props: { items: listItem[] }) => {
	return (
		<Section>
			<div>
				<Div>
					{props.items.map((item) => {
						return (
							<ItemSingle item={item} key={item.id} hLevel={2} />
						)
					})}
				</Div>
			</div>
		</Section>
	)
}

export default ItemList
