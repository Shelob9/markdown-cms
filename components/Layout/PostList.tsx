import React from 'react'

import { listItem } from './ItemList'
import ContentSection from './ContentSection'
import { Heading, P, Div, A, Box } from './primatives'

const Post: React.FC<{
	hLevel?: 1 | 2 | 3 | 4
	post: listItem
	RenderContent?: ({ post: listItem }) => JSX.Element
}> = ({ post, RenderContent, hLevel }) => {
	const { title, content, date, to } = post
	function createMarkup() {
		return { __html: content }
	}

	return (
		<Box as={'article'}>
			<Div
				className="w-full px-4 md:px-6 text-xl text-grey-darkest leading-normal"
				style={{ fontFamily: 'Georgia,serif' }}
			>
				<Div className="font-sans">
					<Heading
						level={hLevel}
						className="font-sans break-normal text-black pt-6 pb-2 text-3xl md:text-4xl"
					>
						<A href={post.to}>{title}</A>
					</Heading>
					{date && (
						<P className="text-sm md:text-base font-normal text-grey-dark">
							{date}
						</P>
					)}
				</Div>
				{RenderContent ? (
					<RenderContent post={post} />
				) : (
					<div
						className="py-6"
						dangerouslySetInnerHTML={createMarkup()}
					/>
				)}
			</Div>
		</Box>
	)
}
const PostList = (props: {
	posts: listItem[]
	title?: string
	hLevel?: 1 | 2 | 3 | 4
	RenderContent?: ({ post: listItem }) => JSX.Element
}) => {
	const { posts, RenderContent } = props
	let hLevel = props.hLevel ? props.hLevel : 1
	return (
		<ContentSection
			hLevel={hLevel}
			title={props.title ? props.title : 'Posts'}
		>
			<>
				{posts.map((post) => (
					<Post
						key={post.title}
						post={post}
						RenderContent={RenderContent}
						//@ts-ignore
						hLevel={hLevel + 1}
					/>
				))}
			</>
		</ContentSection>
	)
}

export default PostList
