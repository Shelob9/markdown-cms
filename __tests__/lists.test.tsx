import * as React from 'react';
import PostList from '../components/Layout/PostList';
import {ItemSingle} from '../components/Layout/ItemList';

import { render } from '@testing-library/react';
describe('PostList', () => {

    test('matches snapshot', () => {
        const { container } = render(<PostList
            title={'The Title'}
            posts={[
                {
                    id: 1,
                    content: 'One',
                    title: 'One Time',
                    to: '/one'
                },
                {
                    id: 2,
                    content: 'Two',
                    title: 'One Two Time',
                    to: '/two'
                },
            ]}
        />);
        expect(
            container
        ).toMatchSnapshot();
   })

})
describe('ItemSingle', () => {
    test('matches snapshot', () => {
        const { container } = render(
            <ItemSingle
                hLevel={2}
                item={{
                    id: 2,
                    content: 'Two',
                    title: 'One Two Time',
                    to: '/two'
                }}
            />
        );
        expect(container).toMatchSnapshot();
    });
})