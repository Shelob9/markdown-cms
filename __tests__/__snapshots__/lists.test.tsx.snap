// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSingle matches snapshot 1`] = `
<div>
  <div>
    <div>
      <h2>
        <a
          href="/two"
        >
          One Two Time
        </a>
      </h2>
      <a
        href="/two"
      >
        Read More
      </a>
    </div>
    <hr />
  </div>
</div>
`;

exports[`PostList matches snapshot 1`] = `
<div>
  <section
    class="text-gray-700 body-font overflow-hidden"
  >
    <div>
      <div />
      <h1
        class="text-2xl font-medium text-gray-900 title-font mb-2"
      >
        The Title
      </h1>
      <article>
        <div
          class="w-full px-4 md:px-6 text-xl text-grey-darkest leading-normal"
          style="font-family: Georgia,serif;"
        >
          <div
            class="font-sans"
          >
            <h2
              class="font-sans break-normal text-black pt-6 pb-2 text-3xl md:text-4xl"
            >
              <a
                href="/one"
              >
                One Time
              </a>
            </h2>
          </div>
          <div
            class="py-6"
          >
            One
          </div>
        </div>
      </article>
      <article>
        <div
          class="w-full px-4 md:px-6 text-xl text-grey-darkest leading-normal"
          style="font-family: Georgia,serif;"
        >
          <div
            class="font-sans"
          >
            <h2
              class="font-sans break-normal text-black pt-6 pb-2 text-3xl md:text-4xl"
            >
              <a
                href="/two"
              >
                One Two Time
              </a>
            </h2>
          </div>
          <div
            class="py-6"
          >
            Two
          </div>
        </div>
      </article>
    </div>
  </section>
</div>
`;
